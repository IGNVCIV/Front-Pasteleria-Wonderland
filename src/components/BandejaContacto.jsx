import { useState, useEffect } from "react";

function BandejaContacto() {
  const [mensajes, setMensajes] = useState([]);
  const [mensajeSeleccionado, setMensajeSeleccionado] = useState(null);

  useEffect(() => {
    fetch("/mensajes.json")
      .then((res) => res.json())
      .then((data) => {
        if (!data || data.length === 0) {
          // mensaje de ejemplo
          setMensajes([
            {
              fecha: new Date().toLocaleDateString(),
              nombre: "Torta de matrimonio",
              correo: "fran@duocuc.cl",
              orden: "-",
              mensaje:
                "Necesitamos una torta de boda muy grande, de tres pisos, con decoración elegante y flores comestibles. El evento será el 15 de diciembre y necesitamos que la entrega sea puntual en la iglesia central. Además, queremos que tenga sabor vainilla con relleno de frambuesa en todos los pisos. Por favor, confirmen disponibilidad y precio."
            }
          ]);
        } else {
          setMensajes(data);
        }
      })
      .catch((err) => {
        console.error("Error cargando mensajes:", err);
        setMensajes([
          {
            fecha: new Date().toLocaleDateString(),
            nombre: "Torta de matrimonio",
            correo: "fran@duocuc.cl",
            orden: "-",
            mensaje:
              "Necesitamos una torta de boda muy grande, de tres pisos, con decoración elegante y flores comestibles. El evento será el 15 de diciembre y necesitamos que la entrega sea puntual en la iglesia central. Además, queremos que tenga sabor vainilla con relleno de frambuesa en todos los pisos. Por favor, confirmen disponibilidad y precio."
          }
        ]);
      });
  }, []);

  return (
    <div>
      <h2 className="h4 mb-3">Bandeja de contacto</h2>
      <p className="text-muted">Mensajes recibidos desde el formulario.</p>

      <div className="card shadow-sm">
        <div className="table-responsive">
          <table className="table table-hover align-middle mb-0">
            <thead className="table-light">
              <tr>
                <th>Fecha</th>
                <th>Nombre</th>
                <th>Correo</th>
                <th>Orden</th>
                <th>Mensaje</th>
                <th>Acción</th>
              </tr>
            </thead>
            <tbody>
              {mensajes.map((m, index) => (
                <tr key={index}>
                  <td>{m.fecha}</td>
                  <td>{m.nombre}</td>
                  <td>{m.correo}</td>
                  <td>{m.orden || "-"}</td>
                  <td>
                    {m.mensaje.length > 50
                      ? m.mensaje.substring(0, 50) + "..."
                      : m.mensaje}
                  </td>
                  <td>
                    <button
                      className="btn btn-sm btn-outline-primary"
                      onClick={() => setMensajeSeleccionado(m)}
                    >
                      Ver completo
                    </button>
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      </div>

      {/* Modal simple */}
      {mensajeSeleccionado && (
        <div
          style={{
            position: "fixed",
            top: 0,
            left: 0,
            width: "100%",
            height: "100%",
            backgroundColor: "rgba(0,0,0,0.5)",
            display: "flex",
            justifyContent: "center",
            alignItems: "center",
            zIndex: 9999,
          }}
          onClick={() => setMensajeSeleccionado(null)}
        >
          <div
            style={{
              backgroundColor: "#fff",
              padding: "1.5rem",
              borderRadius: "8px",
              maxWidth: "600px",
              width: "90%",
              maxHeight: "80%",
              overflowY: "auto",
              boxShadow: "0 0 10px rgba(0,0,0,0.3)",
            }}
            onClick={(e) => e.stopPropagation()}
          >
            <h5>Mensaje de {mensajeSeleccionado.nombre}</h5>
            <p>{mensajeSeleccionado.mensaje}</p>
            <div className="text-end">
              <button
                className="btn btn-secondary btn-sm"
                onClick={() => setMensajeSeleccionado(null)}
              >
                Cerrar
              </button>
            </div>
          </div>
        </div>
      )}
    </div>
  );
}

export default BandejaContacto;
