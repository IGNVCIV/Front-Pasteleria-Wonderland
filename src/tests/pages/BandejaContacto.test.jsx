import '@testing-library/jest-dom';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { describe, it, expect, beforeEach } from 'vitest';
import { MemoryRouter } from 'react-router-dom';
import BandejaContacto from '../../components/BandejaContacto';


const mockMensajes = [
  {
    fecha: '19/10/2025',
    nombre: 'Cliente de prueba',
    correo: 'cliente@test.com',
    orden: 'ORD123',
    mensaje: 'Mensaje de prueba para verificar renderizado.',
  },
];

describe('Componente BandejaContacto', () => {
  beforeEach(() => {
    global.fetch = vi.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve(mockMensajes),
      })
    );
  });

  it('renderiza correctamente el título y descripción', () => {
    render(
      <MemoryRouter>
        <BandejaContacto />
      </MemoryRouter>
    );

    expect(screen.getByText(/Bandeja de contacto/i)).toBeInTheDocument();
    expect(screen.getByText(/Mensajes recibidos desde el formulario/i)).toBeInTheDocument();
  });

  it('muestra los datos del mensaje obtenido del fetch', async () => {
    render(
      <MemoryRouter>
        <BandejaContacto />
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(screen.getByText('Cliente de prueba')).toBeInTheDocument();
      expect(screen.getByText('ORD123')).toBeInTheDocument();
    });
  });

  it('abre el modal al hacer clic en "Ver completo"', async () => {
    render(
      <MemoryRouter>
        <BandejaContacto />
      </MemoryRouter>
    );

    const boton = await screen.findByRole('button', { name: /Ver completo/i });
    fireEvent.click(boton);

    await waitFor(() => {
  const coincidencias = screen.getAllByText(/Mensaje de prueba para verificar renderizado/i);
  expect(coincidencias.length).toBeGreaterThan(1); // o el número exacto si lo sabes
});

  });

  it('cierra el modal al hacer clic en el botón "Cerrar"', async () => {
    render(
      <MemoryRouter>
        <BandejaContacto />
      </MemoryRouter>
    );

    const boton = await screen.findByRole('button', { name: /Ver completo/i });
    fireEvent.click(boton);

    const cerrar = await screen.findByRole('button', { name: /Cerrar/i });
    fireEvent.click(cerrar);

    await waitFor(() => {
      expect(screen.queryByText(/Mensaje de Cliente de prueba/i)).not.toBeInTheDocument();
    });
  });

  it('coincide con el snapshot actual', async () => {
    const { container } = render(
      <MemoryRouter>
        <BandejaContacto />
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});


