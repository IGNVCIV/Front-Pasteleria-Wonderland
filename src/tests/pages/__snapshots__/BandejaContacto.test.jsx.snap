// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Componente BandejaContacto > coincide con el snapshot actual 1`] = `
<div>
  <div>
    <h2
      class="h4 mb-3"
    >
      Bandeja de contacto
    </h2>
    <p
      class="text-muted"
    >
      Mensajes recibidos desde el formulario.
    </p>
    <div
      class="card shadow-sm"
    >
      <div
        class="table-responsive"
      >
        <table
          class="table table-hover align-middle mb-0"
        >
          <thead
            class="table-light"
          >
            <tr>
              <th>
                Fecha
              </th>
              <th>
                Nombre
              </th>
              <th>
                Correo
              </th>
              <th>
                Orden
              </th>
              <th>
                Mensaje
              </th>
              <th>
                Acci√≥n
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </div>
</div>
`;
